### PR TITLE
Show a warning message on suspicious usage of collect_parameters and …

### LINKED ISSUE #47: Show a warning message on suspected usages of parameter GPU migration
`Function.to_gpu` and `FunctionSet.to_gpu` involves array copying, while an Optimizer object does not know this event. Consequently, user must migrate parameters onto devices BEFORE passing them to the optimizer. Violating this rule does not emit any messages currently, which ends up to making the training code meaningless.

So how about show a warning message if `FunctionSet.collect_parameters` is called after `FunctionSet.to_gpu`?

### PR DESCRIPTION
**Pull request by beam2d**
…to_gpu/to_cpu. It fixes #47.

### CODE REVIEW

**Comment by unnonouno** [comment]
Please make a test to check warning.

File: chainer/function_set.py [hunk 1]
```diff
@@ -118,4 +130,14 @@ def gradients(self, grads):
             func.gradients = grad_iter
 
     def _get_sorted_funcs(self):
-        return sorted(self.__dict__.iteritems())
+        return sorted(((k, v) for k, v in self.__dict__.iteritems()
+                       if isinstance(v, Function)))
+
+    def _warn_if_collected(self):
+        if hasattr(self, '__collected'):
```
Comment (unnonouno) [inline chainer/function_set.py]: Do we need to check the value of `__collected`?

**Comment by beam2d** [comment]
I will add tests after #5 is resolved.

File: chainer/function_set.py (hunk 1 above)
Comment (beam2d) [inline chainer/function_set.py]: I used the existence of __collected as a flag.

Comment (unnonouno) [inline chainer/function_set.py]: OK, no problem. Users can't set `__collected` to `False`.

**Comment by beam2d** [comment]
I found that users may want to call `to_cpu` after all training processes are done. However, the code of this PR shows a (non-sense) warning message on such usage. I have no idea for this problem yet.

**Comment by beam2d** [comment]
I cannot come up with a good solution yet, while this patch gets old, so close it. Anyone with a better idea can reopen it or open another issue/PR.

**Pull request closed without merging**